store/postgres/src/mapping.rs

```diff
@@ -0,0 +1,1141 @@
+    pub fn as_str(&self) -> &str {
+        &self.0
+    }
+}
+
+impl From<&str> for SqlName {
+    fn from(name: &str) -> Self {
+        SqlName(name.to_snake_case())
+    }
+}
+
+impl From<String> for SqlName {
+    fn from(name: String) -> Self {
+        SqlName(name.to_snake_case())
+    }
+}
+
+impl fmt::Display for SqlName {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        self.0.fmt(f)
+    }
+}
+
+/// The SQL type to use for GraphQL ID properties. We support
+/// strings and byte arrays
+#[derive(Copy, Clone, Debug, Eq, PartialEq)]
+pub enum IdType {
+    String,
+    #[allow(dead_code)]
+    Bytes,
+}
+
+#[derive(Debug, Clone)]
+pub struct Mapping {
+    /// The SQL type for columns with GraphQL type `ID`
+    id_type: IdType,
+    /// Maps the GraphQL name of a type to the relational table
+    pub tables: HashMap<String, Rc<Table>>,
+    /// A fake table that mirrors the Query type for the schema
+    root_table: Table,
+    /// The subgraph id
+    pub subgraph: SubgraphDeploymentId,
+    /// The database schema for this subgraph
+    pub schema: String,
+    /// Map the entity names of interfaces to the list of
+    /// database tables that contain entities implementing
+    /// that interface
+    pub interfaces: HashMap<String, Vec<Rc<Table>>>,
+    /// The query to count all entities
+    pub count_query: String,
+}
+
+impl Mapping {
+    /// Generate a mapping for a relational schema for entities in the
+    /// GraphQL schema `document`. Attributes of type `ID` will use the
+    /// SQL type `id_type`. The subgraph ID is passed in `subgraph`, and
+    /// the name of the database schema in which the subgraph's tables live
+    /// is in `schema`.
+    pub fn new<V>(
+        document: &s::Document,
+        id_type: IdType,
+        subgraph: SubgraphDeploymentId,
+        schema: V,
+    ) -> Result<Mapping, StoreError>
+    where
+        V: Into<String>,
+    {
+        use s::Definition::*;
+        use s::TypeDefinition::*;
+
+        let schema = schema.into();
+
+        // Check that we can handle all the definitions
+        for defn in &document.definitions {
+            match defn {
+                TypeDefinition(Object(_)) | TypeDefinition(Interface(_)) => (),
+                other => {
+                    return Err(StoreError::Unknown(format_err!(
+                        "can not handle {:?}",
+                        other
+                    )))
+                }
+            }
+        }
+
+        // Extract interfaces and tables
+        let mut interfaces: HashMap<String, Vec<SqlName>> = HashMap::new();
+        let mut tables = Vec::new();
+        let mut position = 0;
+
+        for defn in &document.definitions {
+            match defn {
+                TypeDefinition(Object(obj_type)) => {
+                    position += 1;
+                    let table = Table::new(obj_type, &mut interfaces, id_type, position)?;
+                    tables.push(table);
```

Comment:
Looks like `position` can be replaced by `tables.len()`.

Reply:
Commit 154d276d

Reply:
Out of curiosity, why the preference for 1-based indexing?

Reply:
I think it started by accident with how I manipulated `position`, and leaving it 1-based saved me from fixing up the schema generation tests.

Reply:
Switched to 0-based numbering in commit 424fb8ca